### PR TITLE
Generics validation

### PR DESCRIPTION
**Pull request by andriy-dmytruk**
Adds validation for generics such as `List<@NotBlank String>` or`Map<String, List<@NotBlank String>>`.
The `DefaultValidator` was modified to accommodate for the functionality. Refer to changes in tests for examples and details, like path display in this case.

Additionally, publisher validation was modified to be performed in form of `Mono<@Valid Pojo>`. Validation for single publishers (like `Mono`) was fixed.

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -20,6 +20,7 @@ rootProject.name = 'validation-parent'
 
 include 'validation'
 include 'validation-bom'
+include 'validation-visitor'
```
Comment (graemerocher) [inline settings.gradle:23]: rename this to `validation-processor` for consistency with other models

File: validation-visitor/build.gradle [hunk 2]
```diff
@@ -0,0 +1,34 @@
+plugins {
+    id "io.micronaut.build.internal.validation-module"
+}
+
+repositories {
+    mavenCentral()
+    maven { url "https://s01.oss.sonatype.org/content/repositories/snapshots/" }
+}
```
Comment (graemerocher) [inline validation-visitor/build.gradle:8]: put this in `settings.gradle` under `dependencyManagement`

File: validation-visitor/src/main/java/io/micronaut/validation/visitor/ValidationVisitor.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+public class ValidationVisitor implements TypeElementVisitor<Object, Object> {
+
+    private static final String ANN_CONSTRAINT = "javax.validation.Constraint";
+    private static final String ANN_VALID = "javax.validation.Valid";
+
+    private ClassElement classElement;
+
+    @Override
+    public Set<String> getSupportedAnnotationNames() {
+        return new HashSet<>(Arrays.asList(ANN_CONSTRAINT, ANN_VALID));
```
Comment (graemerocher) [inline validation-visitor/src/main/java/io/micronaut/validation/visitor/ValidationVisitor.java:53]: Use `Set.of`

File: validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java [hunk 4]
```diff
@@ -111,14 +111,15 @@ public Object intercept(MethodInvocationContext<Object, Object> context) {
                     throw new ConstraintViolationException(constraintViolations);
                 }
             }
-            if (hasValidationAnnotation(context)) {
+            // if (hasValidationAnnotation(context)) {
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java:114]: Is this intentional? If so remove the commented out code blocks

File: validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java [hunk 5]
```diff
@@ -2074,17 +1798,51 @@ public <T extends Path.Node> T as(Class<T> nodeType) {
     }
 
     /**
-     * Method node implementation.
+     * Default constructor node.
      */
-    private final class DefaultParameterNode extends DefaultPropertyNode implements Path.ParameterNode {
+    private static final class DefaultConstructorNode extends DefaultMethodNode implements Path.ConstructorNode {
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java:1803]: make this a `record`

File: validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java [hunk 6]
```diff
@@ -2074,17 +1798,51 @@ public <T extends Path.Node> T as(Class<T> nodeType) {
...
+        @Override
+        public ElementKind getKind() {
+            return ElementKind.CONSTRUCTOR;
+        }
+    }
+
+    /**
+     * Default parameter node implementation.
+     */
+    private static final class DefaultParameterNode implements Path.ParameterNode {
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java:1817]: make this a `record`

File: validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java [hunk 7]
```diff
@@ -2159,36 +1901,34 @@ public <T extends Path.Node> T as(Class<T> nodeType) {
     }
 
     /**
-     * Default property node impl.
+     * Default container element node implementation.
      */
-    private class DefaultPropertyNode implements Path.PropertyNode {
+    private static final class DefaultContainerElementNode implements Path.ContainerElementNode {
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java:1906]: make this a record

File: validation/src/main/java/io/micronaut/validation/validator/ReactiveValidator.java [hunk 8]
```diff
@@ -32,12 +33,13 @@ public interface ReactiveValidator {
...
      * constraint violation error occurs a {@link javax.validation.ConstraintViolationException} will be thrown.
      *
+     * @param returnType The required type of publisher
      * @param publisher The publisher
      * @param groups The groups
      * @param <T> The generic type
      * @return The publisher
      */
-    @NonNull <T> Publisher<T> validatePublisher(@NonNull Publisher<T> publisher, Class<?>... groups);
+    @NonNull <T> Publisher<T> validatePublisher(ReturnType returnType, @NonNull Publisher<T> publisher, Class<?>... groups);
```
Comment (graemerocher) [inline validation/src/main/java/io/micronaut/validation/validator/ReactiveValidator.java:42]: add nullability annotation to first parameter

**Review by graemerocher — CHANGES_REQUESTED** [review]

**Review by graemerocher — APPROVED** [review]

File: validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java (hunk 5 above)
Comment (andriy-dmytruk) [reply validation/src/main/java/io/micronaut/validation/validator/DefaultValidator.java:1783]: Noticed that the code size could be reduced by using a common parent, so I created an abstract `DefaultNode` with 5 methods. `record` didn't seem to reduce the implementation, as it doesn't define getters in the desired format.

**Comment by graemerocher** [comment]
@andriy-dmytruk shall I merge this?

**Pull request merged**